keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,560 @@
+      - name: nvidia.com/mig-7g.40gb
+        nominalQuota: 0  # not allowed -- inefficient use of GPU resources
+```
+
+[mig]: https://docs.nvidia.com/datacenter/cloud-native/kubernetes/latest/index.html#the-mixed-strategy
+[instaslice]: https://github.com/openshift/instaslice-operator
+
+#### Story 2
+
+This KEP would allow Kueue to augment a Job's resource requests
+for quota purposes with additional resources that are not actually 
+available in the Nodes of the cluster. For example, it could be used to 
+define quotas in terms of a fictional currency `kueue.x-k8s.io/credits`
+to give "costs" to various resources. This could be used to approximate
+the actual monetary cost of a cloud resource (for example a larger or more powerful GPU
+could cost twice as many `credits` as a smaller or previous generation GPU).
+We'd expect in this usage pattern that `Replace` would be used so that
+resource list is just augmented with the computed `credits` resource.  We'd also
+expect that a ClusterQueue's `borrowingLimit` for `credits` would be set to `0`
+to enforce the intended fiscal constraint.
+
+## Design Details
+
+Since Kueue's workload.Info struct is already a well-defined
+place where resource filtering is being applied to compute the effective resource
+request from a Workload's PodSpec, the core of the implementation is
+straightforward and follows from how `excludedResourcePrefixes` are already processed.
+
+The non-trivial design issues are how to configure the transformations,
+the possible granularity of that configuration, and how to maximize observability
+of the effective resources requested by a Workload.
+
+### Specifying the Transformation
+
+A transformation matrix is defined as a single object.  Most likely
+this object would be embedded inside Kueue's `Configuration` API, but
+it would also be possible to make it a separate top-level config API or 
+use a config map containing a single yaml field that would be deserialized
+into the type defined below.  To support finer-grained configuration, instead
+of defining a single controller-wide transformation matrix, they could be embedded
+in ClusterQueue ResourceGroups (perhaps implicitly by instead embedding the
+transformation matrix in a ResourceFlavor).
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  ConversionBaseValues map[corev1.ResourceName]resource.Quantity `json:"conversionBaseValues"`
+
+  // should the input resource be replaced or retained
+  Strategy ResourceTransformationStrategy `json:"strategy"`
+}
+
+// The map key is the input resource name
+type ResourceTransformationMatrix map[corev1.ResourceName]ResourceTransformation
```

Comment:
Follow API conventions regarding the use of lists over maps https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
The primary concern for the typed map is the typed map is dropped from CRD scheme.
So, as Aldo mentioned above, let's use the slice. The dropping scheme will bring us some limitations to using the K8s functionality like the CEL computation budget.